### ISSUE TITLE
Issue while running ALOpsAppTest

### ISSUE THREAD

**Issue by gintuoklis** [HodorNV/ALOps#54]
Hey Guys, 

We do receive teh following error while runing ALOpsAppTest task:

My container comes witht the database that contains testtoolkit already (old fashioned C/AL v14). Therefore, we do not import testtoolkit.

**##[error]Run-Test-Exception: Exception calling ".ctor" with "2" argument(s): "Value cannot be null. Parameter name: logicalControl"**

Our yaml snippet for running tests
```
- task: ALOpsAppTest@1
  displayName: 'Run TestSuite'
  inputs:
    usedocker: true
    testfilter: '50000..99999'
    fixed_tag: test_tag
    import_testtoolkit: false
    show_available_tests: true
```

Full printout:
```
Starting: Run TestSuite

==============================================================================

Task         : ALOps App Test

Description  : ALOps App Test

Version      : 1.426.933

Author       : Hodor

Help         : Run Business Central Test-Suite and collect results.

==============================================================================

*** Validate configuration

*** Importing required PS-Functions

*** Working folder: C:\agent\_work\2\s

***

*** Prepare Docker Session

*** Connect Docker Session

*** Initiate Docker Session

*** Copy Modules to Docker

*** Setup Docker Session

*** Import Modules in Docker Session

*** Import UI Client [C:\Run\DevOps\BCAppTest\bin\Microsoft.Dynamics.Framework.UI.Client.dll]

VERBOSE: Loading module from path 'C:\Run\DevOps\BCAppTest\bin\Microsoft.Dynamics.Framework.UI.Client.dll'.

*** Import NewtonSoft JSON [C:\Program Files\Microsoft Dynamics NAV\140\Service\Add-ins\LSRetail\Newtonsoft.Json\Newtonsoft.Json.dll]

*** Import Modules in Docker Session: class_ClientContext

*** Setup BC Credentials in Docker

*** Set TCP Keep-Alive

*** Start TestSuite

*** Running TestSuite with authentication [NavUserPassword] with User [***]

*** Run warm-up with page [130409]

*** Warm-up Client Context

*** Create Client Context

*** Add Tests to Default-TestSuite with filter [50000..99999]

Apply filter for Test-Codeunits: [50000..99999]

##[error]Run-Test-Exception: Exception calling ".ctor" with "2" argument(s): "Value cannot be null.

Parameter name: logicalControl"

*** Get XUnit Test file from [] to [C:\agent\_work\2\s\]

##[error]Cannot bind argument to parameter 'Path' because it is null.

Finishing: Run TestSuite

```

**Comment by AdminHodor** [comment]
Hello @gintuoklis ,

For BC v14 the ImportTestToolkit is required, the default one included in v14 is not capable of setting the number ranges to filter. Modification where made to the default objects, which are included in ALOps and installed in the TestStep with the "import_testtoolkit" parameter.
The error you are getting refers to a Control that is missing on the page to set the number range filter.

BC v15 does have all the required controls by default and does not require modifications.

We uploaded the specific FOB on GitHub, if you do not want to use the "ImportTestkit" option you can import the FOB in your "template database". [BC v14 Modified Page](https://github.com/HodorNV/ALOps/blob/AdminHodor/FOB/pstesttoolpage_auto.fob).

What we could also do is a provide an extra parameter that would only import the modified page and not the entire test-suite, which would save some time on the step. 

Kind regards,

Bart Bourgeois

**Comment by gintuoklis** [comment]
Thanks for the swift reply. 
Indeed I used standard test tool that comes with _navcontainerhelper_. After installing yours it all worked fine.

I see yours tool have advantage over the standard one from _navcontainerhelper_. There you were forced to add installation codeunit for each extension. With yours it's no longer needed.

I just added one small modification to your function GetTestCodeunits, to filter out all other C/AL test codeunits that we still have in core application. 

`FromAllObjWithCaption.SETFILTER("App Package ID",  '<>%1', '00000000-0000-0000-0000-000000000000');
`

**Comment by AdminHodor** [comment]
Hello @gintuoklis ,

>There you were forced to add installation codeunit for each extension

The interesting thing is, for v15 we made decided to follow the standard, the more we can avoid 'custom' modifications the better. So for v15, we support running Test-Suites by name, which does require adding "Install-Codeunits" to create them. We struggled with some partner in the conversion, but after the facts we have to admit it's a better approach and better manageable. We follow the guidance of Freddy Kristiansen close in this matter.

Never-the-less, we are always open to suggestions, serving the community. If requested by the community we will adapt ALOps to support the v14 scenario's in v15. 

Glad we could be of assistance!

Kind regards,

Bart Bourgeois
